// @adobe/alloy/libEs5/components/Personalization/createComponent.js
"use strict";

exports.default = void 0;
var _index = require("../../utils/index.js");
var _createPersonalizationDetails = require("./createPersonalizationDetails.js");
var _loggerMessage = require("./constants/loggerMessage.js");
var _propositionEventType = require("../../constants/propositionEventType.js");
var _validateApplyPropositionsOptions = require("./validateApplyPropositionsOptions.js");
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
var _default = exports.default = function _default(_ref) {
  var getPageLocation = _ref.getPageLocation,
    logger = _ref.logger,
    fetchDataHandler = _ref.fetchDataHandler,
    viewChangeHandler = _ref.viewChangeHandler,
    onClickHandler = _ref.onClickHandler,
    isAuthoringModeEnabled = _ref.isAuthoringModeEnabled,
    mergeQuery = _ref.mergeQuery,
    viewCache = _ref.viewCache,
    showContainers = _ref.showContainers,
    applyPropositions = _ref.applyPropositions,
    setTargetMigration = _ref.setTargetMigration,
    mergeDecisionsMeta = _ref.mergeDecisionsMeta,
    renderedPropositions = _ref.renderedPropositions,
    onDecisionHandler = _ref.onDecisionHandler,
    handleConsentFlicker = _ref.handleConsentFlicker;
  return {
    lifecycle: {
      onComponentsRegistered: function onComponentsRegistered() {
        handleConsentFlicker();
      },
      onDecision: onDecisionHandler,
      onBeforeRequest: function onBeforeRequest(_ref2) {
        var request = _ref2.request;
        setTargetMigration(request);
        return Promise.resolve();
      },
      onBeforeEvent: function onBeforeEvent(_ref3) {
        var event = _ref3.event,
          renderDecisions = _ref3.renderDecisions,
          _ref3$decisionScopes = _ref3.decisionScopes,
          decisionScopes = _ref3$decisionScopes === void 0 ? [] : _ref3$decisionScopes,
          _ref3$personalization = _ref3.personalization,
          personalization = _ref3$personalization === void 0 ? {} : _ref3$personalization,
          _ref3$onResponse = _ref3.onResponse,
          onResponse = _ref3$onResponse === void 0 ? _index.noop : _ref3$onResponse,
          _ref3$onRequestFailur = _ref3.onRequestFailure,
          onRequestFailure = _ref3$onRequestFailur === void 0 ? _index.noop : _ref3$onRequestFailur;
        // Include propositions on all responses, overridden with data as needed
        onResponse(function () {
          return {
            propositions: []
          };
        });
        onRequestFailure(function () {
          return showContainers();
        });
        if (isAuthoringModeEnabled()) {
          logger.warn(_loggerMessage.AUTHORING_ENABLED);

          // If we are in authoring mode we disable personalization
          mergeQuery(event, {
            enabled: false
          });
          return Promise.resolve();
        }
        var personalizationDetails = (0, _createPersonalizationDetails.default)({
          getPageLocation: getPageLocation,
          renderDecisions: renderDecisions,
          decisionScopes: decisionScopes,
          personalization: personalization,
          event: event,
          isCacheInitialized: viewCache.isInitialized(),
          logger: logger
        });
        var decisionsMetaPromises = [];
        if (personalizationDetails.shouldIncludeRenderedPropositions()) {
          decisionsMetaPromises.push(renderedPropositions.clear());
        }
        if (personalizationDetails.shouldFetchData()) {
          var cacheUpdate = viewCache.createCacheUpdate(personalizationDetails.getViewName());
          onRequestFailure(function () {
            return cacheUpdate.cancel();
          });
          fetchDataHandler({
            cacheUpdate: cacheUpdate,
            personalizationDetails: personalizationDetails,
            event: event,
            onResponse: onResponse
          });
        } else if (personalizationDetails.shouldUseCachedData()) {
          // eslint-disable-next-line consistent-return
          decisionsMetaPromises.push(viewChangeHandler({
            personalizationDetails: personalizationDetails,
            event: event,
            onResponse: onResponse,
            onRequestFailure: onRequestFailure
          }));
        }

        // This promise.all waits for both the pending display notifications to be resolved
        // (i.e. the top of page call to finish rendering) and the view change handler to
        // finish rendering anything for this view.
        return Promise.all(decisionsMetaPromises).then(function (decisionsMetas) {
          // We only want to call mergeDecisionsMeta once, but we can get the propositions
          // from two places: the pending display notifications and the view change handler.
          var decisionsMeta = (0, _index.flatMap)(decisionsMetas, function (dms) {
            return dms;
          });
          if ((0, _index.isNonEmptyArray)(decisionsMeta)) {
            mergeDecisionsMeta(event, decisionsMeta, [_propositionEventType.PropositionEventType.DISPLAY]);
          }
        });
      },
      onClick: function onClick(_ref4) {
        var event = _ref4.event,
          clickedElement = _ref4.clickedElement;
        onClickHandler({
          event: event,
          clickedElement: clickedElement
        });
      }
    },
    commands: {
      applyPropositions: {
        optionsValidator: function optionsValidator(options) {
          return (0, _validateApplyPropositionsOptions.default)({
            logger: logger,
            options: options
          });
        },
        run: applyPropositions
      }
    }
  };
};
