// @adobe/alloy/libEs5/components/Personalization/index.js
"use strict";

exports.default = void 0;
var _index = require("../../utils/validation/index.js");
var _createComponent = require("./createComponent.js");
var _index2 = require("./dom-actions/index.js");
var _createCollect = require("./createCollect.js");
var _index3 = require("./flicker/index.js");
var _createFetchDataHandler = require("./createFetchDataHandler.js");
var _collectClicks = require("./dom-actions/clicks/collectClicks.js");
var _isAuthoringModeEnabled = require("./utils/isAuthoringModeEnabled.js");
var _event = require("./event.js");
var _createOnClickHandler = require("./createOnClickHandler.js");
var _createViewCacheManager = require("./createViewCacheManager.js");
var _createViewChangeHandler = require("./createViewChangeHandler.js");
var _createClickStorage2 = require("./createClickStorage.js");
var _createApplyPropositions = require("./createApplyPropositions.js");
var _createGetPageLocation = require("./createGetPageLocation.js");
var _createSetTargetMigration = require("./createSetTargetMigration.js");
var _remapCustomCodeOffers = require("./dom-actions/remapCustomCodeOffers.js");
var _remapHeadOffers = require("./dom-actions/remapHeadOffers.js");
var _createPreprocess = require("./dom-actions/createPreprocess.js");
var _injectCreateProposition = require("./handlers/injectCreateProposition.js");
var _createAsyncArray = require("./utils/createAsyncArray.js");
var schema = require("../../constants/schema.js");
var _processDefaultContent = require("./handlers/processDefaultContent.js");
var _surfaceUtils = require("./utils/surfaceUtils.js");
var _createProcessDomAction = require("./handlers/createProcessDomAction.js");
var _createProcessHtmlContent = require("./handlers/createProcessHtmlContent.js");
var _createProcessRedirect = require("./handlers/createProcessRedirect.js");
var _createProcessPropositions = require("./handlers/createProcessPropositions.js");
var _createOnDecisionHandler = require("./createOnDecisionHandler.js");
var _createProcessInAppMessage = require("./handlers/createProcessInAppMessage.js");
var _initInAppMessageActionsModules = require("./in-app-message-actions/initInAppMessageActionsModules.js");
var _createRedirect = require("./dom-actions/createRedirect.js");
var _createNotificationHandler = require("./createNotificationHandler.js");
var _createHandleConsentFlicker = require("./createHandleConsentFlicker.js");
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); } /*
Copyright 2019 Adobe. Ackll rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
var createPersonalization = function createPersonalization(_ref) {
  var config = _ref.config,
    logger = _ref.logger,
    eventManager = _ref.eventManager,
    consent = _ref.consent;
  var targetMigrationEnabled = config.targetMigrationEnabled,
    prehidingStyle = config.prehidingStyle;
  var collect = (0, _createCollect.default)({
    eventManager: eventManager,
    mergeDecisionsMeta: _event.mergeDecisionsMeta
  });
  var _createClickStorage = (0, _createClickStorage2.default)(),
    getClickMetasBySelector = _createClickStorage.getClickMetasBySelector,
    getClickSelectors = _createClickStorage.getClickSelectors,
    storeClickMetrics = _createClickStorage.storeClickMetrics;
  var getPageLocation = (0, _createGetPageLocation.default)({
    window: window
  });
  var domActionsModules = (0, _index2.initDomActionsModules)();
  var preprocess = (0, _createPreprocess.default)([_remapHeadOffers.default, _remapCustomCodeOffers.default]);
  var createProposition = (0, _injectCreateProposition.default)({
    preprocess: preprocess,
    isPageWideSurface: _surfaceUtils.isPageWideSurface
  });
  var viewCache = (0, _createViewCacheManager.default)({
    createProposition: createProposition
  });
  var executeRedirect = (0, _createRedirect.default)(window);
  var schemaProcessors = _defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty({}, schema.DEFAULT_CONTENT_ITEM, _processDefaultContent.default), schema.DOM_ACTION, (0, _createProcessDomAction.default)({
    modules: domActionsModules,
    logger: logger,
    storeClickMetrics: storeClickMetrics
  })), schema.HTML_CONTENT_ITEM, (0, _createProcessHtmlContent.default)({
    modules: domActionsModules,
    logger: logger
  })), schema.REDIRECT_ITEM, (0, _createProcessRedirect.default)({
    logger: logger,
    executeRedirect: executeRedirect,
    collect: collect
  })), schema.MESSAGE_IN_APP, (0, _createProcessInAppMessage.default)({
    modules: (0, _initInAppMessageActionsModules.default)(collect),
    logger: logger
  }));
  var processPropositions = (0, _createProcessPropositions.default)({
    schemaProcessors: schemaProcessors,
    logger: logger
  });
  var renderedPropositions = (0, _createAsyncArray.default)();
  var notificationHandler = (0, _createNotificationHandler.default)(collect, renderedPropositions);
  var fetchDataHandler = (0, _createFetchDataHandler.default)({
    prehidingStyle: prehidingStyle,
    showContainers: _index3.showContainers,
    hideContainers: _index3.hideContainers,
    mergeQuery: _event.mergeQuery,
    processPropositions: processPropositions,
    createProposition: createProposition,
    notificationHandler: notificationHandler,
    consent: consent
  });
  var onClickHandler = (0, _createOnClickHandler.default)({
    mergeDecisionsMeta: _event.mergeDecisionsMeta,
    collectClicks: _collectClicks.default,
    getClickSelectors: getClickSelectors,
    getClickMetasBySelector: getClickMetasBySelector
  });
  var viewChangeHandler = (0, _createViewChangeHandler.default)({
    processPropositions: processPropositions,
    viewCache: viewCache
  });
  var applyPropositions = (0, _createApplyPropositions.default)({
    processPropositions: processPropositions,
    createProposition: createProposition,
    renderedPropositions: renderedPropositions,
    viewCache: viewCache
  });
  var setTargetMigration = (0, _createSetTargetMigration.default)({
    targetMigrationEnabled: targetMigrationEnabled
  });
  var onDecisionHandler = (0, _createOnDecisionHandler.default)({
    processPropositions: processPropositions,
    createProposition: createProposition,
    notificationHandler: notificationHandler
  });
  var handleConsentFlicker = (0, _createHandleConsentFlicker.default)({
    showContainers: _index3.showContainers,
    consent: consent
  });
  return (0, _createComponent.default)({
    getPageLocation: getPageLocation,
    logger: logger,
    fetchDataHandler: fetchDataHandler,
    viewChangeHandler: viewChangeHandler,
    onClickHandler: onClickHandler,
    isAuthoringModeEnabled: _isAuthoringModeEnabled.default,
    mergeQuery: _event.mergeQuery,
    viewCache: viewCache,
    showContainers: _index3.showContainers,
    applyPropositions: applyPropositions,
    setTargetMigration: setTargetMigration,
    mergeDecisionsMeta: _event.mergeDecisionsMeta,
    renderedPropositions: renderedPropositions,
    onDecisionHandler: onDecisionHandler,
    handleConsentFlicker: handleConsentFlicker
  });
};
createPersonalization.namespace = "Personalization";
createPersonalization.configValidators = (0, _index.objectOf)({
  prehidingStyle: (0, _index.string)().nonEmpty(),
  targetMigrationEnabled: (0, _index.boolean)().default(false)
});
var _default = exports.default = createPersonalization;
