// @adobe/alloy/libEs5/core/createEvent.js
"use strict";

exports.default = void 0;
var _index = require("../utils/index.js");
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }
function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; } /*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
var getXdmPropositions = function getXdmPropositions(xdm) {
  return xdm &&
  // eslint-disable-next-line no-underscore-dangle
  xdm._experience &&
  // eslint-disable-next-line no-underscore-dangle
  xdm._experience.decisioning &&
  // eslint-disable-next-line no-underscore-dangle
  (0, _index.isNonEmptyArray)(xdm._experience.decisioning.propositions) ?
  // eslint-disable-next-line no-underscore-dangle
  xdm._experience.decisioning.propositions : [];
};
var _default = exports.default = function _default() {
  var content = {};
  var userXdm;
  var userData;
  var _documentMayUnload = false;
  var isFinalized = false;
  var shouldSendEvent = true;
  var throwIfEventFinalized = function throwIfEventFinalized(methodName) {
    if (isFinalized) {
      throw new Error(methodName + " cannot be called after event is finalized.");
    }
  };
  var event = {
    hasQuery: function hasQuery() {
      return Object.prototype.hasOwnProperty.call(this.getContent(), "query");
    },
    getContent: function getContent() {
      var currentContent = JSON.parse(JSON.stringify(content));
      if (userXdm) {
        (0, _index.deepAssign)(currentContent, {
          xdm: userXdm
        });
      }
      if (userData) {
        (0, _index.deepAssign)(currentContent, {
          data: userData
        });
      }
      return currentContent;
    },
    setUserXdm: function setUserXdm(value) {
      throwIfEventFinalized("setUserXdm");
      userXdm = value;
    },
    setUserData: function setUserData(value) {
      throwIfEventFinalized("setUserData");
      userData = value;
    },
    mergeXdm: function mergeXdm(xdm) {
      throwIfEventFinalized("mergeXdm");
      if (xdm) {
        (0, _index.deepAssign)(content, {
          xdm: xdm
        });
      }
    },
    mergeData: function mergeData(data) {
      throwIfEventFinalized("mergeData");
      if (data) {
        (0, _index.deepAssign)(content, {
          data: data
        });
      }
    },
    mergeMeta: function mergeMeta(meta) {
      throwIfEventFinalized("mergeMeta");
      if (meta) {
        (0, _index.deepAssign)(content, {
          meta: meta
        });
      }
    },
    mergeQuery: function mergeQuery(query) {
      throwIfEventFinalized("mergeQuery");
      if (query) {
        (0, _index.deepAssign)(content, {
          query: query
        });
      }
    },
    documentMayUnload: function documentMayUnload() {
      _documentMayUnload = true;
    },
    finalize: function finalize(onBeforeEventSend) {
      if (isFinalized) {
        return;
      }
      var newPropositions = (0, _index.deduplicateArray)([].concat(_toConsumableArray(getXdmPropositions(userXdm)), _toConsumableArray(getXdmPropositions(content.xdm))), function (a, b) {
        return a === b || a.id && b.id && a.id === b.id && a.scope && b.scope && a.scope === b.scope;
      });
      if (userXdm) {
        this.mergeXdm(userXdm);
      }
      if (newPropositions.length > 0) {
        // eslint-disable-next-line no-underscore-dangle
        content.xdm._experience.decisioning.propositions = newPropositions;
      }
      if (userData) {
        event.mergeData(userData);
      }

      // the event should already be considered finalized in case onBeforeEventSend throws an error
      isFinalized = true;
      if (onBeforeEventSend) {
        // assume that the onBeforeEventSend callback will fail (in-case of an error)
        shouldSendEvent = false;

        // this allows the user to replace the xdm and data properties
        // on the object passed to the callback
        var tempContent = {
          xdm: content.xdm || {},
          data: content.data || {}
        };
        var result = onBeforeEventSend(tempContent);
        shouldSendEvent = result !== false;
        content.xdm = tempContent.xdm || {};
        content.data = tempContent.data || {};
        if ((0, _index.isEmptyObject)(content.xdm)) {
          delete content.xdm;
        }
        if ((0, _index.isEmptyObject)(content.data)) {
          delete content.data;
        }
      }
    },
    getDocumentMayUnload: function getDocumentMayUnload() {
      return _documentMayUnload;
    },
    isEmpty: function isEmpty() {
      return (0, _index.isEmptyObject)(content) && (!userXdm || (0, _index.isEmptyObject)(userXdm)) && (!userData || (0, _index.isEmptyObject)(userData));
    },
    shouldSend: function shouldSend() {
      return shouldSendEvent;
    },
    getViewName: function getViewName() {
      if (!userXdm || !userXdm.web || !userXdm.web.webPageDetails) {
        return undefined;
      }
      return userXdm.web.webPageDetails.viewName;
    },
    toJSON: function toJSON() {
      if (!isFinalized) {
        throw new Error("toJSON called before finalize");
      }
      return content;
    }
  };
  return event;
};
