// @adobe/alloy/libEs5/components/Identity/index.js
"use strict";

exports.default = void 0;
var _index = require("../../utils/index.js");
var _injectProcessIdSyncs = require("./injectProcessIdSyncs.js");
var _configValidators = require("./configValidators.js");
var _createComponent = require("./createComponent.js");
var _createLegacyIdentity = require("./createLegacyIdentity.js");
var _awaitVisitorOptIn = require("./visitorService/awaitVisitorOptIn.js");
var _injectGetEcidFromVisitor = require("./visitorService/injectGetEcidFromVisitor.js");
var _injectHandleResponseForIdSyncs = require("./injectHandleResponseForIdSyncs.js");
var _injectEnsureSingleIdentity = require("./injectEnsureSingleIdentity.js");
var _addEcidQueryToPayload = require("./addEcidQueryToPayload.js");
var _injectSetDomainForInitialIdentityPayload = require("./injectSetDomainForInitialIdentityPayload.js");
var _injectAddLegacyEcidToPayload = require("./injectAddLegacyEcidToPayload.js");
var _injectAddQueryStringIdentityToPayload = require("./injectAddQueryStringIdentityToPayload.js");
var _addEcidToPayload = require("./addEcidToPayload.js");
var _injectAwaitIdentityCookie = require("./injectAwaitIdentityCookie.js");
var _getEcidFromResponse = require("./getEcidFromResponse.js");
var _createGetIdentity = require("./getIdentity/createGetIdentity.js");
var _createIdentityRequest = require("./getIdentity/createIdentityRequest.js");
var _createIdentityRequestPayload = require("./getIdentity/createIdentityRequestPayload.js");
var _injectAppendIdentityToUrl = require("./appendIdentityToUrl/injectAppendIdentityToUrl.js");
/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

var createIdentity = function createIdentity(_ref) {
  var config = _ref.config,
    logger = _ref.logger,
    consent = _ref.consent,
    fireReferrerHideableImage = _ref.fireReferrerHideableImage,
    sendEdgeNetworkRequest = _ref.sendEdgeNetworkRequest,
    apexDomain = _ref.apexDomain;
  var orgId = config.orgId,
    thirdPartyCookiesEnabled = config.thirdPartyCookiesEnabled,
    globalConfigOverrides = config.edgeConfigOverrides;
  var getEcidFromVisitor = (0, _injectGetEcidFromVisitor.default)({
    logger: logger,
    orgId: orgId,
    awaitVisitorOptIn: _awaitVisitorOptIn.default
  });
  var loggingCookieJar = (0, _index.createLoggingCookieJar)({
    logger: logger,
    cookieJar: _index.cookieJar
  });
  var legacyIdentity = (0, _createLegacyIdentity.default)({
    config: config,
    getEcidFromVisitor: getEcidFromVisitor,
    apexDomain: apexDomain,
    cookieJar: loggingCookieJar,
    isPageSsl: window.location.protocol === "https:"
  });
  var doesIdentityCookieExist = (0, _index.injectDoesIdentityCookieExist)({
    orgId: orgId
  });
  var getIdentity = (0, _createGetIdentity.default)({
    sendEdgeNetworkRequest: sendEdgeNetworkRequest,
    createIdentityRequestPayload: _createIdentityRequestPayload.default,
    createIdentityRequest: _createIdentityRequest.default,
    globalConfigOverrides: globalConfigOverrides
  });
  var setDomainForInitialIdentityPayload = (0, _injectSetDomainForInitialIdentityPayload.default)({
    thirdPartyCookiesEnabled: thirdPartyCookiesEnabled,
    areThirdPartyCookiesSupportedByDefault: _index.areThirdPartyCookiesSupportedByDefault
  });
  var addLegacyEcidToPayload = (0, _injectAddLegacyEcidToPayload.default)({
    getLegacyEcid: legacyIdentity.getEcid,
    addEcidToPayload: _addEcidToPayload.default
  });
  var addQueryStringIdentityToPayload = (0, _injectAddQueryStringIdentityToPayload.default)({
    locationSearch: window.document.location.search,
    dateProvider: function dateProvider() {
      return new Date();
    },
    orgId: orgId,
    logger: logger
  });
  var awaitIdentityCookie = (0, _injectAwaitIdentityCookie.default)({
    doesIdentityCookieExist: doesIdentityCookieExist,
    orgId: orgId,
    logger: logger
  });
  var ensureSingleIdentity = (0, _injectEnsureSingleIdentity.default)({
    doesIdentityCookieExist: doesIdentityCookieExist,
    setDomainForInitialIdentityPayload: setDomainForInitialIdentityPayload,
    addLegacyEcidToPayload: addLegacyEcidToPayload,
    awaitIdentityCookie: awaitIdentityCookie,
    logger: logger
  });
  var processIdSyncs = (0, _injectProcessIdSyncs.default)({
    fireReferrerHideableImage: fireReferrerHideableImage,
    logger: logger
  });
  var handleResponseForIdSyncs = (0, _injectHandleResponseForIdSyncs.default)({
    processIdSyncs: processIdSyncs
  });
  var appendIdentityToUrl = (0, _injectAppendIdentityToUrl.default)({
    dateProvider: function dateProvider() {
      return new Date();
    },
    orgId: orgId,
    globalConfigOverrides: globalConfigOverrides
  });
  return (0, _createComponent.default)({
    addEcidQueryToPayload: _addEcidQueryToPayload.default,
    addQueryStringIdentityToPayload: addQueryStringIdentityToPayload,
    ensureSingleIdentity: ensureSingleIdentity,
    setLegacyEcid: legacyIdentity.setEcid,
    handleResponseForIdSyncs: handleResponseForIdSyncs,
    getEcidFromResponse: _getEcidFromResponse.default,
    getIdentity: getIdentity,
    consent: consent,
    appendIdentityToUrl: appendIdentityToUrl,
    logger: logger,
    config: config
  });
};
createIdentity.namespace = "Identity";
createIdentity.configValidators = _configValidators.default;
var _default = exports.default = createIdentity;
