// @adobe/alloy/libEs5/components/Personalization/flicker/index.js
"use strict";

exports.showElements = exports.showContainers = exports.hideElements = exports.hideContainers = void 0;
var _index = require("../../../utils/dom/index.js");
var _tagName = require("../../../constants/tagName.js");
var _index2 = require("../dom-actions/dom/index.js");
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); } /*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
var PREHIDING_ID = "alloy-prehiding";
var HIDING_STYLE_DEFINITION = "{ visibility: hidden }";

// Using global is OK since we have a single DOM
// so storing nodes even for multiple Alloy instances is fine
var styleNodes = {};
var hideElements = exports.hideElements = function hideElements(prehidingSelector) {
  // if we have different events with the same
  // prehiding selector we don't want to recreate
  // the style tag
  if (styleNodes[prehidingSelector]) {
    return;
  }
  var nonce = (0, _index2.getNonce)();
  var attrs = _objectSpread({}, nonce && {
    nonce: nonce
  });
  var props = {
    textContent: prehidingSelector + " " + HIDING_STYLE_DEFINITION
  };
  var node = (0, _index.createNode)(_tagName.STYLE, attrs, props);
  (0, _index.appendNode)(document.head, node);
  styleNodes[prehidingSelector] = node;
};
var showElements = exports.showElements = function showElements(prehidingSelector) {
  var node = styleNodes[prehidingSelector];
  if (node) {
    (0, _index.removeNode)(node);
    delete styleNodes[prehidingSelector];
  }
};
var hideContainers = exports.hideContainers = function hideContainers(prehidingStyle) {
  if (!prehidingStyle) {
    return;
  }

  // If containers prehiding style has been added
  // by customer's prehiding snippet we don't
  // want to add the same node
  var node = (0, _index2.getElementById)(PREHIDING_ID);
  if (node) {
    return;
  }
  var nonce = (0, _index2.getNonce)();
  var attrs = _objectSpread({
    id: PREHIDING_ID
  }, nonce && {
    nonce: nonce
  });
  var props = {
    textContent: prehidingStyle
  };
  var styleNode = (0, _index.createNode)(_tagName.STYLE, attrs, props);
  (0, _index.appendNode)(document.head, styleNode);
};
var showContainers = exports.showContainers = function showContainers() {
  // If containers prehiding style exists
  // we will remove it
  var node = (0, _index2.getElementById)(PREHIDING_ID);
  if (!node) {
    return;
  }
  (0, _index.removeNode)(node);
};
