// @adobe/alloy/libEs5/components/DecisioningEngine/createEvaluateRulesetsCommand.js
"use strict";

exports.default = void 0;
var _index = require("../../utils/validation/index.js");
/*
Copyright 2023 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

var validateOptions = function validateOptions(_ref) {
  var options = _ref.options;
  var validator = (0, _index.objectOf)({
    renderDecisions: (0, _index.boolean)(),
    personalization: (0, _index.objectOf)({
      decisionContext: (0, _index.objectOf)({})
    })
  }).noUnknownFields();
  return validator(options);
};
var _default = exports.default = function _default(_ref2) {
  var contextProvider = _ref2.contextProvider,
    decisionProvider = _ref2.decisionProvider;
  var run = function run(_ref3) {
    var renderDecisions = _ref3.renderDecisions,
      decisionContext = _ref3.decisionContext,
      applyResponse = _ref3.applyResponse;
    return applyResponse({
      renderDecisions: renderDecisions,
      propositions: decisionProvider.evaluate(contextProvider.getContext(decisionContext))
    });
  };
  var optionsValidator = function optionsValidator(options) {
    return validateOptions({
      options: options
    });
  };
  return {
    optionsValidator: optionsValidator,
    run: run
  };
};
