// @adobe/alloy/libEs5/components/StreamingMedia/createTrackMediaEvent.js
"use strict";

exports.default = void 0;
var _eventTypes = require("./constants/eventTypes.js");
/*
Copyright 2024 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

var getContentState = function getContentState(eventType, sessionContentState) {
  if (eventType === _eventTypes.default.AD_START || eventType === _eventTypes.default.Ad_BREAK_START || eventType === _eventTypes.default.AD_SKIP || eventType === _eventTypes.default.AD_COMPLETE) {
    return "ad";
  }
  if (eventType === _eventTypes.default.AD_BREAK_COMPLETE || eventType === _eventTypes.default.CHAPTER_COMPLETE || eventType === _eventTypes.default.CHAPTER_START || eventType === _eventTypes.default.CHAPTER_SKIP || eventType === _eventTypes.default.SESSION_START) {
    return "main";
  }
  if (eventType === _eventTypes.default.SESSION_END || eventType === _eventTypes.default.SESSION_COMPLETE) {
    return "completed";
  }
  return sessionContentState;
};
var _default = exports.default = function _default(_ref) {
  var mediaEventManager = _ref.mediaEventManager,
    mediaSessionCacheManager = _ref.mediaSessionCacheManager,
    config = _ref.config;
  var sendMediaEvent = function sendMediaEvent(options) {
    var event = mediaEventManager.createMediaEvent({
      options: options
    });
    var playerId = options.playerId,
      xdm = options.xdm;
    var eventType = xdm.eventType;
    var action = eventType.split(".")[1];
    var _mediaSessionCacheMan = mediaSessionCacheManager.getSession(playerId),
      getPlayerDetails = _mediaSessionCacheMan.getPlayerDetails,
      sessionPromise = _mediaSessionCacheMan.sessionPromise,
      playbackState = _mediaSessionCacheMan.playbackState;
    return sessionPromise.then(function (result) {
      if (!result.sessionId) {
        return Promise.reject(new Error("Failed to trigger media event: " + eventType + ". Session ID is not available for playerId: " + playerId + "."));
      }
      mediaEventManager.augmentMediaEvent({
        event: event,
        eventType: eventType,
        playerId: playerId,
        getPlayerDetails: getPlayerDetails,
        sessionID: result.sessionId
      });
      return mediaEventManager.trackMediaEvent({
        event: event,
        action: action
      }).then(function () {
        if (playerId) {
          if (eventType === _eventTypes.default.SESSION_COMPLETE || eventType === _eventTypes.default.SESSION_END) {
            mediaSessionCacheManager.stopPing({
              playerId: playerId
            });
          } else {
            var sessionPlaybackState = getContentState(eventType, playbackState);
            if (sessionPlaybackState === "completed") {
              return;
            }
            var interval = sessionPlaybackState === "ad" ? config.streamingMedia.adPingInterval : config.streamingMedia.mainPingInterval;
            var pingId = setTimeout(function () {
              var pingOptions = {
                playerId: playerId,
                xdm: {
                  eventType: _eventTypes.default.PING
                }
              };
              sendMediaEvent(pingOptions);
            }, interval * 1000);
            mediaSessionCacheManager.savePing({
              playerId: playerId,
              pingId: pingId,
              playbackState: sessionPlaybackState
            });
          }
        }
      });
    });
  };
  return function (options) {
    return sendMediaEvent(options);
  };
};
