// @adobe/alloy/libEs5/components/StreamingMedia/createTrackMediaSession.js
"use strict";

exports.default = void 0;
var _playbackState = require("./constants/playbackState.js");
/*
Copyright 2024 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
var _default = exports.default = function _default(_ref) {
  var config = _ref.config,
    mediaEventManager = _ref.mediaEventManager,
    mediaSessionCacheManager = _ref.mediaSessionCacheManager,
    _ref$legacy = _ref.legacy,
    legacy = _ref$legacy === void 0 ? false : _ref$legacy;
  return function (options) {
    if (!config.streamingMedia) {
      return Promise.reject(new Error("Streaming media is not configured."));
    }
    var playerId = options.playerId,
      getPlayerDetails = options.getPlayerDetails;
    var event = mediaEventManager.createMediaSession(options);
    mediaEventManager.augmentMediaEvent({
      event: event,
      playerId: playerId,
      getPlayerDetails: getPlayerDetails
    });
    var sessionPromise = mediaEventManager.trackMediaSession({
      event: event,
      mediaOptions: {
        playerId: playerId,
        getPlayerDetails: getPlayerDetails,
        legacy: legacy
      }
    });
    mediaSessionCacheManager.storeSession({
      playerId: playerId,
      sessionDetails: {
        sessionPromise: sessionPromise,
        getPlayerDetails: getPlayerDetails,
        playbackState: _playbackState.default.MAIN
      }
    });
    return sessionPromise;
  };
};
