// @adobe/alloy/libEs5/components/StreamingMedia/createStreamingMediaComponent.js
"use strict";

exports.default = void 0;
var _index = require("../../utils/index.js");
var _validateMediaSessionOptions = require("./validateMediaSessionOptions.js");
var _validateMediaEventOptions = require("./validateMediaEventOptions.js");
/*
Copyright 2024 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
var _default = exports.default = function _default(_ref) {
  var config = _ref.config,
    trackMediaEvent = _ref.trackMediaEvent,
    trackMediaSession = _ref.trackMediaSession,
    mediaResponseHandler = _ref.mediaResponseHandler;
  return {
    lifecycle: {
      onBeforeEvent: function onBeforeEvent(_ref2) {
        var mediaOptions = _ref2.mediaOptions,
          _ref2$onResponse = _ref2.onResponse,
          onResponse = _ref2$onResponse === void 0 ? _index.noop : _ref2$onResponse;
        if (!mediaOptions) {
          return;
        }
        var legacy = mediaOptions.legacy,
          playerId = mediaOptions.playerId,
          getPlayerDetails = mediaOptions.getPlayerDetails;
        if (legacy) {
          return;
        }
        onResponse(function (_ref3) {
          var response = _ref3.response;
          return mediaResponseHandler({
            playerId: playerId,
            getPlayerDetails: getPlayerDetails,
            response: response
          });
        });
      }
    },
    commands: {
      createMediaSession: {
        optionsValidator: function optionsValidator(options) {
          return (0, _validateMediaSessionOptions.default)({
            options: options
          });
        },
        run: trackMediaSession
      },
      sendMediaEvent: {
        optionsValidator: function optionsValidator(options) {
          return (0, _validateMediaEventOptions.default)({
            options: options
          });
        },
        run: function run(options) {
          if (!config.streamingMedia) {
            return Promise.reject(new Error("Streaming media is not configured."));
          }
          return trackMediaEvent(options);
        }
      }
    }
  };
};
