// @adobe/alloy/libEs5/core/edgeNetwork/injectApplyResponse.js
"use strict";

exports.default = void 0;
var _index = require("../../utils/index.js");
var _mergeLifecycleResponses = require("./mergeLifecycleResponses.js");
var _handleRequestFailure = require("./handleRequestFailure.js");
/*
Copyright 2023 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

var HTTP_STATUS_OK = 200;
var _default = exports.default = function _default(_ref) {
  var cookieTransfer = _ref.cookieTransfer,
    lifecycle = _ref.lifecycle,
    createResponse = _ref.createResponse,
    processWarningsAndErrors = _ref.processWarningsAndErrors;
  return function (_ref2) {
    var request = _ref2.request,
      responseHeaders = _ref2.responseHeaders,
      responseBody = _ref2.responseBody,
      _ref2$runOnResponseCa = _ref2.runOnResponseCallbacks,
      runOnResponseCallbacks = _ref2$runOnResponseCa === void 0 ? _index.noop : _ref2$runOnResponseCa,
      _ref2$runOnRequestFai = _ref2.runOnRequestFailureCallbacks,
      runOnRequestFailureCallbacks = _ref2$runOnRequestFai === void 0 ? _index.noop : _ref2$runOnRequestFai;
    var onResponseCallbackAggregator = (0, _index.createCallbackAggregator)();
    onResponseCallbackAggregator.add(lifecycle.onResponse);
    onResponseCallbackAggregator.add(runOnResponseCallbacks);
    var onRequestFailureCallbackAggregator = (0, _index.createCallbackAggregator)();
    onRequestFailureCallbackAggregator.add(lifecycle.onRequestFailure);
    onRequestFailureCallbackAggregator.add(runOnRequestFailureCallbacks);
    var getHeader = function getHeader(key) {
      return responseHeaders[key];
    };
    return lifecycle.onBeforeRequest({
      request: request,
      onResponse: onResponseCallbackAggregator.add,
      onRequestFailure: onRequestFailureCallbackAggregator.add
    }).then(function () {
      return processWarningsAndErrors({
        statusCode: HTTP_STATUS_OK,
        getHeader: getHeader,
        body: JSON.stringify(responseBody),
        parsedBody: responseBody
      });
    }).catch((0, _handleRequestFailure.default)(onRequestFailureCallbackAggregator)).then(function () {
      var response = createResponse({
        content: responseBody,
        getHeader: getHeader
      });

      // This will clobber any cookies set via HTTP from the server.  So care should be given to remove any state:store handles if that is not desirable
      cookieTransfer.responseToCookies(response);
      return onResponseCallbackAggregator.call({
        response: response
      }).then(_mergeLifecycleResponses.default);
    });
  };
};
