// @adobe/alloy/libEs5/components/ActivityCollector/createGetLinkDetails.js
"use strict";

exports.default = void 0;
/*
Copyright 2022 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
var _default = exports.default = function _default(_ref) {
  var window = _ref.window,
    getLinkName = _ref.getLinkName,
    getLinkRegion = _ref.getLinkRegion,
    getAbsoluteUrlFromAnchorElement = _ref.getAbsoluteUrlFromAnchorElement,
    findSupportedAnchorElement = _ref.findSupportedAnchorElement,
    determineLinkType = _ref.determineLinkType;
  return function (_ref2) {
    var targetElement = _ref2.targetElement,
      config = _ref2.config,
      logger = _ref2.logger;
    var anchorElement = findSupportedAnchorElement(targetElement);
    if (!anchorElement) {
      logger.info("This link click event is not triggered because the HTML element is not an anchor.");
      return undefined;
    }
    var linkUrl = getAbsoluteUrlFromAnchorElement(window, anchorElement);
    if (!linkUrl) {
      logger.info("This link click event is not triggered because the HTML element doesn't have an URL.");
      return undefined;
    }
    var linkType = determineLinkType(window, config, linkUrl, anchorElement);
    var linkRegion = getLinkRegion(anchorElement);
    var linkName = getLinkName(anchorElement);
    var onBeforeLinkClickSend = config.onBeforeLinkClickSend;
    var options = {
      xdm: {
        eventType: "web.webinteraction.linkClicks",
        web: {
          webInteraction: {
            name: linkName,
            region: linkRegion,
            type: linkType,
            URL: linkUrl,
            linkClicks: {
              value: 1
            }
          }
        }
      },
      data: {},
      clickedElement: targetElement
    };
    if (!onBeforeLinkClickSend) {
      return options;
    }
    var shouldEventBeTracked = onBeforeLinkClickSend(options);
    if (shouldEventBeTracked !== false) {
      return options;
    }
    logger.info("This link click event is not triggered because it was canceled in onBeforeLinkClickSend.");
    return undefined;
  };
};
