// @adobe/alloy/libEs5/components/ActivityCollector/index.js
"use strict";

exports.default = void 0;
var _attachClickActivityCollector = require("./attachClickActivityCollector.js");
var _configValidators = require("./configValidators.js");
var _createLinkClick = require("./createLinkClick.js");
var _createGetLinkDetails = require("./createGetLinkDetails.js");
var _getLinkName = require("./getLinkName.js");
var _getLinkRegion = require("./getLinkRegion.js");
var _utils = require("./utils.js");
/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

var _getLinkDetails = (0, _createGetLinkDetails.default)({
  window: window,
  getLinkName: _getLinkName.default,
  getLinkRegion: _getLinkRegion.default,
  getAbsoluteUrlFromAnchorElement: _utils.getAbsoluteUrlFromAnchorElement,
  findSupportedAnchorElement: _utils.findSupportedAnchorElement,
  determineLinkType: _utils.determineLinkType
});
var createActivityCollector = function createActivityCollector(_ref) {
  var config = _ref.config,
    eventManager = _ref.eventManager,
    handleError = _ref.handleError,
    logger = _ref.logger;
  var linkClick = (0, _createLinkClick.default)({
    getLinkDetails: _getLinkDetails,
    config: config,
    logger: logger
  });
  return {
    lifecycle: {
      onComponentsRegistered: function onComponentsRegistered(tools) {
        var lifecycle = tools.lifecycle;
        (0, _attachClickActivityCollector.default)({
          eventManager: eventManager,
          lifecycle: lifecycle,
          handleError: handleError
        });
        // TODO: createScrollActivityCollector ...
      },
      onClick: function onClick(_ref2) {
        var event = _ref2.event,
          clickedElement = _ref2.clickedElement;
        linkClick({
          targetElement: clickedElement,
          event: event
        });
      }
    }
  };
};
createActivityCollector.namespace = "ActivityCollector";
createActivityCollector.configValidators = _configValidators.default;
createActivityCollector.buildOnInstanceConfiguredExtraParams = function (_ref3) {
  var config = _ref3.config,
    logger = _ref3.logger;
  return {
    getLinkDetails: function getLinkDetails(targetElement) {
      return _getLinkDetails({
        targetElement: targetElement,
        config: config,
        logger: logger
      });
    }
  };
};
var _default = exports.default = createActivityCollector;
