// @adobe/alloy/libEs5/components/LegacyMediaAnalytics/index.js
"use strict";

exports.default = void 0;
var _createMediaEventManager = require("../StreamingMedia/createMediaEventManager.js");
var _createMediaSessionCacheManager = require("../StreamingMedia/createMediaSessionCacheManager.js");
var _createTrackMediaEvent = require("../StreamingMedia/createTrackMediaEvent.js");
var _createTrackMediaSession = require("../StreamingMedia/createTrackMediaSession.js");
var _createMediaResponseHandler = require("../StreamingMedia/createMediaResponseHandler.js");
var _createLegacyMediaComponent = require("./createLegacyMediaComponent.js");
var _createMediaHelper = require("./createMediaHelper.js");
var _createGetInstance = require("./createGetInstance.js");
var _injectTimestamp = require("../Context/injectTimestamp.js");
/*
Copyright 2024 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
/* eslint-disable import/no-restricted-paths */

var createLegacyMediaAnalytics = function createLegacyMediaAnalytics(_ref) {
  var eventManager = _ref.eventManager,
    sendEdgeNetworkRequest = _ref.sendEdgeNetworkRequest,
    config = _ref.config,
    logger = _ref.logger,
    consent = _ref.consent;
  var mediaSessionCacheManager = (0, _createMediaSessionCacheManager.default)({
    config: config
  });
  var mediaEventManager = (0, _createMediaEventManager.default)({
    sendEdgeNetworkRequest: sendEdgeNetworkRequest,
    config: config,
    consent: consent,
    eventManager: eventManager,
    setTimestamp: (0, _injectTimestamp.default)(function () {
      return new Date();
    })
  });
  var trackMediaEvent = (0, _createTrackMediaEvent.default)({
    mediaSessionCacheManager: mediaSessionCacheManager,
    mediaEventManager: mediaEventManager,
    config: config
  });
  var trackMediaSession = (0, _createTrackMediaSession.default)({
    config: config,
    mediaEventManager: mediaEventManager,
    mediaSessionCacheManager: mediaSessionCacheManager,
    legacy: true
  });
  var mediaResponseHandler = (0, _createMediaResponseHandler.default)({
    mediaSessionCacheManager: mediaSessionCacheManager,
    config: config,
    trackMediaEvent: trackMediaEvent
  });
  return (0, _createLegacyMediaComponent.default)({
    mediaResponseHandler: mediaResponseHandler,
    trackMediaSession: trackMediaSession,
    trackMediaEvent: trackMediaEvent,
    createMediaHelper: _createMediaHelper.default,
    createGetInstance: _createGetInstance.default,
    logger: logger,
    config: config
  });
};
createLegacyMediaAnalytics.namespace = "Legacy Media Analytics";
var _default = exports.default = createLegacyMediaAnalytics;
