// @adobe/alloy/libEs5/components/Audiences/index.js
"use strict";

exports.default = void 0;
var _index = require("../../utils/index.js");
var _injectProcessDestinations = require("./injectProcessDestinations.js");
var _injectProcessResponse = require("./injectProcessResponse.js");
/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

var createAudiences = function createAudiences(_ref) {
  var logger = _ref.logger,
    fireReferrerHideableImage = _ref.fireReferrerHideableImage;
  // we override the js-cookie converter to encode the cookie value similar on how it is in DIL (PDCL-10238)
  var cookieJarWithEncoding = _index.cookieJar.withConverter({
    write: function write(value) {
      return encodeURIComponent(value);
    }
  });
  var loggingCookieJar = (0, _index.createLoggingCookieJar)({
    logger: logger,
    cookieJar: cookieJarWithEncoding
  });
  var processDestinations = (0, _injectProcessDestinations.default)({
    fireReferrerHideableImage: fireReferrerHideableImage,
    logger: logger,
    cookieJar: loggingCookieJar,
    isPageSsl: window.location.protocol === "https:"
  });
  var processResponse = (0, _injectProcessResponse.default)({
    processDestinations: processDestinations
  });
  return {
    lifecycle: {
      onResponse: processResponse
    },
    commands: {}
  };
};
createAudiences.namespace = "Audiences";
var _default = exports.default = createAudiences;
