// @adobe/alloy/libEs5/components/DataCollector/index.js
"use strict";

exports.default = void 0;
var _validateUserEventOptions = require("./validateUserEventOptions.js");
var _validateApplyResponse = require("./validateApplyResponse.js");
var _index = require("../../utils/index.js");
var _excluded = ["xdm", "data", "documentUnloading", "type", "mergeId", "datasetId", "edgeConfigOverrides"];
/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } } return target; }
var createDataCollector = function createDataCollector(_ref) {
  var eventManager = _ref.eventManager,
    logger = _ref.logger;
  return {
    commands: {
      sendEvent: {
        documentationUri: "https://adobe.ly/3GQ3Q7t",
        optionsValidator: function optionsValidator(options) {
          return (0, _validateUserEventOptions.default)({
            options: options
          });
        },
        run: function run(options) {
          var xdm = options.xdm,
            data = options.data,
            documentUnloading = options.documentUnloading,
            type = options.type,
            mergeId = options.mergeId,
            datasetId = options.datasetId,
            edgeConfigOverrides = options.edgeConfigOverrides,
            eventManagerOptions = _objectWithoutProperties(options, _excluded);
          var event = eventManager.createEvent();
          if (documentUnloading) {
            event.documentMayUnload();
          }
          event.setUserXdm(xdm);
          event.setUserData(data);
          if (type) {
            event.mergeXdm({
              eventType: type
            });
          }
          if (mergeId) {
            event.mergeXdm({
              eventMergeId: mergeId
            });
          }
          if (edgeConfigOverrides) {
            eventManagerOptions.edgeConfigOverrides = edgeConfigOverrides;
          }
          if (datasetId) {
            logger.warn("The 'datasetId' option has been deprecated. Please use 'edgeConfigOverrides.com_adobe_experience_platform.datasets.event.datasetId' instead.");
            eventManagerOptions.edgeConfigOverrides = edgeConfigOverrides || {};
            (0, _index.deepAssign)(eventManagerOptions.edgeConfigOverrides, {
              com_adobe_experience_platform: {
                datasets: {
                  event: {
                    datasetId: datasetId
                  }
                }
              }
            });
          }
          return eventManager.sendEvent(event, eventManagerOptions);
        }
      },
      applyResponse: {
        documentationUri: "",
        optionsValidator: function optionsValidator(options) {
          return (0, _validateApplyResponse.default)({
            options: options
          });
        },
        run: function run(options) {
          var _options$renderDecisi = options.renderDecisions,
            renderDecisions = _options$renderDecisi === void 0 ? false : _options$renderDecisi,
            _options$decisionCont = options.decisionContext,
            decisionContext = _options$decisionCont === void 0 ? {} : _options$decisionCont,
            _options$responseHead = options.responseHeaders,
            responseHeaders = _options$responseHead === void 0 ? {} : _options$responseHead,
            _options$responseBody = options.responseBody,
            responseBody = _options$responseBody === void 0 ? {
              handle: []
            } : _options$responseBody,
            personalization = options.personalization;
          var event = eventManager.createEvent();
          return eventManager.applyResponse(event, {
            renderDecisions: renderDecisions,
            decisionContext: decisionContext,
            responseHeaders: responseHeaders,
            responseBody: responseBody,
            personalization: personalization
          });
        }
      }
    }
  };
};
createDataCollector.namespace = "DataCollector";
var _default = exports.default = createDataCollector;
