// @adobe/alloy/libEs5/constants/schema.js
"use strict";

exports.RULESET_ITEM = exports.REDIRECT_ITEM = exports.MESSAGE_NATIVE_ALERT = exports.MESSAGE_IN_APP = exports.MESSAGE_FEED_ITEM = exports.MEASUREMENT_SCHEMA = exports.JSON_CONTENT_ITEM = exports.HTML_CONTENT_ITEM = exports.DOM_ACTION = exports.DEFAULT_CONTENT_ITEM = void 0;
/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

var DEFAULT_CONTENT_ITEM = exports.DEFAULT_CONTENT_ITEM = "https://ns.adobe.com/personalization/default-content-item";
var DOM_ACTION = exports.DOM_ACTION = "https://ns.adobe.com/personalization/dom-action";
var HTML_CONTENT_ITEM = exports.HTML_CONTENT_ITEM = "https://ns.adobe.com/personalization/html-content-item";
var JSON_CONTENT_ITEM = exports.JSON_CONTENT_ITEM = "https://ns.adobe.com/personalization/json-content-item";
var RULESET_ITEM = exports.RULESET_ITEM = "https://ns.adobe.com/personalization/ruleset-item";
var REDIRECT_ITEM = exports.REDIRECT_ITEM = "https://ns.adobe.com/personalization/redirect-item";
var MESSAGE_IN_APP = exports.MESSAGE_IN_APP = "https://ns.adobe.com/personalization/message/in-app";
var MESSAGE_FEED_ITEM = exports.MESSAGE_FEED_ITEM = "https://ns.adobe.com/personalization/message/feed-item";
var MESSAGE_NATIVE_ALERT = exports.MESSAGE_NATIVE_ALERT = "https://ns.adobe.com/personalization/message/native-alert";
var MEASUREMENT_SCHEMA = exports.MEASUREMENT_SCHEMA = "https://ns.adobe.com/personalization/measurement";
