// @adobe/alloy/libEs5/utils/injectFireReferrerHideableImage.js
"use strict";

exports.default = void 0;
var _fireImage = require("./fireImage.js");
var _index = require("./dom/index.js");
var _tagName = require("../constants/tagName.js");
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; } /*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
var IFRAME_ATTRS = {
  name: "Adobe Alloy"
};
var IFRAME_PROPS = {
  style: {
    display: "none",
    width: 0,
    height: 0
  }
};
var _default = exports.default = function _default() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
    _ref$appendNode = _ref.appendNode,
    appendNode = _ref$appendNode === void 0 ? _index.appendNode : _ref$appendNode,
    _ref$awaitSelector = _ref.awaitSelector,
    awaitSelector = _ref$awaitSelector === void 0 ? _index.awaitSelector : _ref$awaitSelector,
    _ref$createNode = _ref.createNode,
    createNode = _ref$createNode === void 0 ? _index.createNode : _ref$createNode,
    _ref$fireImage = _ref.fireImage,
    fireImage = _ref$fireImage === void 0 ? _fireImage.default : _ref$fireImage;
  var fireOnPage = fireImage;
  var hiddenIframe;
  var createIframe = function createIframe() {
    return awaitSelector(_tagName.BODY).then(function (_ref2) {
      var _ref3 = _slicedToArray(_ref2, 1),
        body = _ref3[0];
      if (hiddenIframe) {
        return hiddenIframe;
      }
      hiddenIframe = createNode(_tagName.IFRAME, IFRAME_ATTRS, IFRAME_PROPS);
      return appendNode(body, hiddenIframe);
    });
  };
  var fireInIframe = function fireInIframe(_ref4) {
    var src = _ref4.src;
    return createIframe().then(function (iframe) {
      var currentDocument = iframe.contentWindow.document;
      return fireImage({
        src: src,
        currentDocument: currentDocument
      });
    });
  };
  return function (request) {
    var hideReferrer = request.hideReferrer,
      url = request.url;
    return hideReferrer ? fireInIframe({
      src: url
    }) : fireOnPage({
      src: url
    });
  };
};
